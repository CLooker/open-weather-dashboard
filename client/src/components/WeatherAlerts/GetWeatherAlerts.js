import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const GetWeatherAlertsWrapper = styled.div`
  margin: 0 5px;
  padding: 0 0 0 10px;
  position: relative;
  bottom: 18px;
`;

export default class GetWeatherAlerts extends Component {
  static propTypes = {
    incrementTriggeredAlertsTotal: PropTypes.func.isRequired,
    registeredAlertsTotal: PropTypes.number,
    setRegisteredAlertsTotal: PropTypes.func.isRequired,
    triggeredAlertsTotal: PropTypes.number
  };

  componentDidMount() {
    const {
      incrementTriggeredAlertsTotal,
      setRegisteredAlertsTotal
    } = this.props;

    fetch(`api/alerts`)
      .then(res => res.json())
      .then(res => {
        setRegisteredAlertsTotal(res.length);

        res.forEach(({ alerts }) => {
          if (Object.keys(alerts).length) incrementTriggeredAlertsTotal(alerts);
        });
      })
      .catch(console.err);
  }

  render() {
    const { registeredAlertsTotal, triggeredAlertsTotal } = this.props;
    return (
      <GetWeatherAlertsWrapper>
        <h6>Registered Alerts: {registeredAlertsTotal}</h6>
        <h6>Triggered Alerts: {triggeredAlertsTotal}</h6>
      </GetWeatherAlertsWrapper>
    );
  }
}
