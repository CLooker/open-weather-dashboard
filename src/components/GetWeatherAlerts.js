import React, { Component } from 'react';
import apiKey from '../api';
import PropTypes from 'prop-types';

export default class GetWeatherAlerts extends Component {
  static propTypes = {
    registeredAlerts: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    triggeredAlerts: PropTypes.array.isRequired,
    updateRegisteredAlerts: PropTypes.func.isRequired,
    updateTriggeredAlerts: PropTypes.func.isRequired
  };

  // place on the prototype
  // so we can spy that it's invoked
  // by componentDidMount
  fetchAlerts() {
    const { updateRegisteredAlerts, updateTriggeredAlerts } = this.props;

    fetch(`http://api.openweathermap.org/data/3.0/triggers?&APPID=${apiKey}`)
      .then(res => res.json())
      .then(res => {
        updateRegisteredAlerts(res);

        res.forEach(({ alerts }) => {
          if (Object.keys(alerts).length > 0) {
            updateTriggeredAlerts(alerts);
          }
        });
      })
      .catch(err => console.log(err));
  }

  componentDidMount() {
    this.fetchAlerts();
  }

  render() {
    return (
      <div className="get-weather-alerts">
        <h6>Registered Alerts: {this.props.registeredAlerts}</h6>
        <h6>Triggered Alerts: {this.props.triggeredAlerts.length}</h6>
      </div>
    );
  }
}
