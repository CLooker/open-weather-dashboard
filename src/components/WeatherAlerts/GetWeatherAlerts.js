import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { apiKey } from '../../utils';

export default class GetWeatherAlerts extends Component {
  static propTypes = {
    incrementTriggeredAlertsTotal: PropTypes.func.isRequired,
    registeredAlertsTotal: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number
    ]),
    setRegisteredAlertsTotal: PropTypes.func.isRequired,
    triggeredAlertsTotal: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number
    ])
  };

  componentDidMount() {
    const {
      incrementTriggeredAlertsTotal,
      setRegisteredAlertsTotal
    } = this.props;

    fetch(`https://api.openweathermap.org/data/3.0/triggers?&APPID=${apiKey}`)
      .then(res => res.json())
      .then(res => {
        setRegisteredAlertsTotal(res.length);

        res.forEach(({ alerts }) => {
          if (Object.keys(alerts).length) incrementTriggeredAlertsTotal(alerts);
        });
      })
      .catch(console.err);
  }

  render() {
    const { registeredAlertsTotal, triggeredAlertsTotal } = this.props;
    return (
      <div className='get-weather-alerts'>
        <h6>Registered Alerts: {registeredAlertsTotal}</h6>
        <h6>Triggered Alerts: {triggeredAlertsTotal}</h6>
      </div>
    );
  }
}
