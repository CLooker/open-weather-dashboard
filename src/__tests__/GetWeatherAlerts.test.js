import React from 'react';
import { mount, shallow } from 'enzyme';
import GetWeatherAlerts from '../components/GetWeatherAlerts';

// refactor fetchAlerts so more testable

describe('GetWeatherAlerts', () => {
  const props = {
    registeredAlerts: 0,
    triggeredAlerts: [],
    updateRegisteredAlerts: jest.fn(),
    updateTriggeredAlerts: jest.fn()
  };

  let component;

  beforeEach(() => (component = shallow(<GetWeatherAlerts {...props} />)));

  afterEach(() => jest.restoreAllMocks());

  describe('initialization', () => {
    it('should render correctly', () => {
      expect(component.debug()).toMatchSnapshot();
    });

    it('should have certain props', () => {
      const actual = component.instance().props;
      const expected = props;
      expect(actual).toEqual(props);
    });

    describe('rendering', () => {
      test('that certain props are rendered', () => {
        const actual = component.html();
        const { registeredAlerts, triggeredAlerts } = props;
        const expected = [registeredAlerts, triggeredAlerts];
        expected.forEach(prop => {
          expect(actual).toContain(prop);
        });
      });

      describe('`componentDidMount`', () => {
        it('should call `fetchAlerts`', () => {
          const spy = jest.spyOn(GetWeatherAlerts.prototype, 'fetchAlerts');
          const mountedComponent = mount(<GetWeatherAlerts {...props} />);
          expect(spy).toHaveBeenCalled();
        });
      });
    });
  });
});
