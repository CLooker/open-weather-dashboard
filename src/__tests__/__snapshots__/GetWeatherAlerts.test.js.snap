// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GetWeatherAlerts initialization should render correctly 1`] = `
"<div className=\\"get-weather-alerts\\">
  <h6>
    Registered Alerts: 
    0
  </h6>
  <h6>
    Triggered Alerts: 
    0
  </h6>
</div>"
`;
